app/controllers/imports_controller.rb

```diff
@@ -0,0 +1,142 @@
+    unless validate_save_models(new_models)
+      aggregate_errors(all_models)
+      return redirect_to root_url
+    end
+
+    # Part 2
+    # Because these following models are dependent on the id's and creation of the previous models,
+    # they have to be parsed, validated and saved afterwards.
+    begin
+      indicators = Indicator.parse(spreadsheet, competencies, levels)
+      resources = Resource.parse(spreadsheet, paradigms)
+    rescue => exception
+      rollback_saved_models(new_models)
+      flash[:error] = EXCEPTION_ERROR_MSG
+      return redirect_to root_url
+    end
+    dependent_models = [indicators, resources]
+
+    # Same behavior as before, except since previous models were saved, 
+    # this will rollback and destroy the created models if there is something invalid.
+    unless validate_save_models(dependent_models)
+      rollback_saved_models(new_models)
+      aggregate_errors(dependent_models)
+      return redirect_to root_url
+    end
+
+    # Part 3
+    # This begins the third level of dependencies for many to many association tables
+    begin
+      indicator_resources = IndicatorResource.parse(spreadsheet, indicators, resources)
+      indicator_questions = IndicatorQuestion.parse(spreadsheet, indicators, questions)
+    rescue => exception
+      rollback_saved_models(new_models + dependent_models)
+      flash[:error] = EXCEPTION_ERROR_MSG
+      return redirect_to root_url
+    end
+    dependent_models_2 = [indicator_resources, indicator_questions]
+
+    # Validate the new models and roll back the previous 2 levels of models.
+    unless validate_save_models(dependent_models_2)
+      rollback_saved_models(new_models + dependent_models)
+      aggregate_errors(dependent_models_2)
+      return redirect_to root_url
+    end
+
+    # Redirect to the original page with success message if no errors have occurred in any of the models
+    flash[:notice] = "Successfully uploaded and imported the #{file.original_filename} spreadsheet."
+    redirect_to root_url
+  end
+
+
+  private
+
+  # Takes in the list of all models that need to be saved into the database.
+  # Checks if the new objects are valid, before saving into the databse.
+  # Nothing is saved unless everything is valid.
+  #
+  # Note: Stops after the first sheet that encounters and error, therefore, error messages
+  # will only be for the first sheet that has errors, not error messages for all sheets.
+  def validate_save_models(models_list)
+    # Goes through all the models and see if its valid and returns false if any of them are invalid
+    models_list.each do |models|
+      unless models.map(&:valid?).all?
+        return false
+      end
+    end
+    # If all are valid, then save them and return true
+    models_list.each do |models|
+      models.each(&:save!)
+    end
+    return true
+  end
+
+  # This method is called if an error exists in the list of models.

```

Comment:
Maybe add into the comment what the different errors could be. Blank row, index out of bounds etc. Probably not completely necessary but might be nice for future teams.